example/shortest_path.cpp

```diff
@@ -0,0 +1,164 @@
+#include <random>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+using namespace std::string_literals;
+
+auto intersperse = [](flux::sequence auto seq, std::string sep) -> flux::generator<std::string>
+{
+    bool first = true;
+    for (auto c: seq) {
+        if (first) {
+            co_yield c;
+            first = false;
+        } else {
+            co_yield sep;
+            co_yield c;
+        }
+    }
+};
+
+namespace color {
+    std::string yellow(const std::string& s) { 
+        return "\u001b[33m"s + s + "\u001b[37m"; 
+    }
+}
+
+class maze {
+private:
+    constexpr static auto wall = std::numeric_limits<int>::max();
+    constexpr static auto path = 0;
+    std::size_t width_;
+    std::size_t height_;
+    std::vector<int> fields_;
+
+    auto adjacent(std::size_t pos) {
+        auto pos_x = pos % width_;
+        auto pos_y = pos / width_;
+        std::vector<std::size_t> adj;
+
+        if (pos_x > 0 && fields_[pos - 1] != wall) { adj.push_back(pos - 1); }
+        if (pos_x < width_ - 1 && fields_[pos + 1] != wall) { adj.push_back(pos + 1); }
+        if (pos_y > 0 && fields_[pos - width_] != wall) { adj.push_back(pos - width_); }
+        if (pos_y < height_ - 1 && fields_[pos + width_] != wall) { adj.push_back(pos + width_); }
+
+        return flux::from(std::move(adj));
+    };
+
+public:
+    maze(std::size_t width = 10, std::size_t height = 10):
+        width_(width),
+        height_(height),
+        fields_(width * height, 0)
+    { 
+        assert(width > 1 && height > 1);
+    }
+
+    static maze random(std::size_t width = 10, std::size_t height = 10) {
+        auto seed = std::random_device{}();
+        std::mt19937 rng(seed);
+        std::uniform_int_distribution dist(1, 9);
+        maze m(width, height);
+        
+        for (auto i: flux::ints(1, m.fields_.size() - 1)) {
+             m.fields_[i] = (rng() % 4) == 0 ? wall : dist(rng);
+        }
+
+        return m;
+    }
+
+    void print(std::ostream& s, bool print_costs = true) const {
+        auto to_char = [print_costs] (int num) {
+            if (num == wall) { return "#"s; }
+            if (num == path) { return color::yellow("*"); }
+            return print_costs ? std::to_string(num) : " "s; 
+        };
+        auto width = width_ * 2 + 3;
+        auto edge = flux::single("|"s);
+        auto h_edge = "+"s + std::string(width - 2, '-') + "+\n"s;
+        auto out = std::ostream_iterator<std::string>(s);
+        
+        s << h_edge;
+        for (auto&& row: flux::ref(fields_).map(to_char).chunk(width_)) {
+            intersperse(flux::chain(flux::ref(edge), FLUX_FWD(row), flux::ref(edge)), " "s).output_to(out);
+            s << '\n';
+        }
+        s << h_edge;
+    }
+
+    void mark_shortest_path() {
+        struct edge_t { std::size_t src, dst; };
+
+        std::vector<std::optional<int>> costs(fields_.size());
+        std::vector<std::optional<size_t>> prevs(fields_.size());
```

Comment:
Using `std::optional` is fine, but could you check whether these work with `flux::optional` as well? Because if not then it means there are bugs I need to fix! :)

Reply:
I changed it to our `flux::optional` but it seems that there is a small difference between standard. I had to use `emplace` instead of `operator=`. Maybe we should fix this.

Reply:
That makes sense. I think it's safer and `emplace` it doesn't look that bad.